src/providers/tpm/utils.rs

```diff
@@ -93,27 +95,64 @@ pub fn to_response_status(error: Error) -> ResponseStatus {
 // The PasswordContext is what is stored by the Key Info Manager.
 #[derive(Serialize, Deserialize, Zeroize)]
 pub struct PasswordContext {
+    /// This value is kept for legacy purposes, to aid in the migration process
+    pub context: TpmsContext,
+    /// This value is confidential and needs to be zeroized by its new owner.
+    pub auth_value: Vec<u8>,
+    pub key_material: KeyMaterial,
+}
+
+// LegacyPasswordContext that stored key contexts only.
+#[derive(Serialize, Deserialize, Zeroize)]
+pub struct LegacyPasswordContext {
     pub context: TpmsContext,
     /// This value is confidential and needs to be zeroized by its new owner.
     pub auth_value: Vec<u8>,
 }
 
+pub fn empty_tpms_ctx() -> TpmsContext {
+    TPMS_CONTEXT::default().try_into().unwrap() // the default value is guaranteed to work
+}
+
 pub fn parsec_to_tpm_params(attributes: Attributes) -> Result<KeyParams> {
     match attributes.key_type {
-        Type::RsaKeyPair => {
+        Type::RsaKeyPair | Type::RsaPublicKey => {
             let size = match attributes.bits {
-                x @ 1024 | x @ 2048 | x @ 3072 | x @ 4096 => x.try_into().unwrap(), // will not fail on the matched values
-                _ => return Err(ResponseStatus::PsaErrorInvalidArgument),
+                1024 => RsaKeyBits::Rsa1024,
+                2048 => RsaKeyBits::Rsa2048,
+                3072 => RsaKeyBits::Rsa3072,
+                4096 => RsaKeyBits::Rsa4096,
+                _ => {
+                    error!(
+                        "Requested RSA key size is not supported ({})",
+                        attributes.bits
+                    );
+                    return Err(ResponseStatus::PsaErrorInvalidArgument);
+                }
```

Comment:
Not sure if it can make use of `try_from()` within the `tss-esapi` crate [here](https://github.com/parallaxsecond/rust-tss-esapi/blob/main/tss-esapi/src/interface_types/key_bits.rs#L134). I would say the conversion should probably lie within the `tss-esapi` crate. Maybe an implementation using `num::FromPrimitive` or just normal match arms.

Reply:
Yeah, there is a `TryFrom<u16>` implemented already, I'll switch to using that

Reply:
Done